File: PLF/labs/lab1/main.py
## 3/a

# Mathematical Model

# listContains(l1l2...ln, el):
#   - false, n = 0
#   - true, l1 = el
#   - listContains(l2...ln, el), otherwise

# checkSet(l1l2...ln):
#   - true, n = 0
#   - false, listContains(l2...ln, l1) = true
#   - checkSet(l2...ln), otherwise
def getLength(list, size = 0):
    if list == []: return size
    return getLength(list[1:], size + 1)

def listContains(list, el):
    if getLength(list) == 0:
        return False
    if list[0] == el:
        return True
    return listContains(list[1:], el)

def checkSet(list):
    if getLength(list) == 0:
        return True
    if listContains(list[1:], list[0]) == True:
        return False
    return checkSet(list[1:])

# getLength (l1l2...ln)
#   - 0, n = 0
#   - 1 + getLength(l2...ln), otherwise

# getLengthMain(l1l2...ln)
#   - getLength(l1l2...ln, 0)

# distinctElements(l1l2...ln, k1k2...kn, length):
#   - length - getLength(k1k2...kn), getLength(l1l2...ln) = 0
#   - distinctElements(l2...ln, l1 U k1k2...kn, length + 1), listContains(l2l3...ln, l1) or listContains(k1k2...kn, l1)
#   - distinctElements(l2...ln, k1k2...kn, length + 1), otherwise

# distinctElementsMain(l1l2...ln):
#   - distinctElements(l1l2...ln, [], 0)


def distinctElements(list, repeatedElements = [], length = 0):
    if getLength(list) == 0:
        return length - getLength(repeatedElements)
    if listContains(list[1:], list[0]) or listContains(repeatedElements, list[0]):
        return distinctElements(list[1:], repeatedElements + [list[0]], length + 1)
    else:
        return distinctElements(list[1:], repeatedElements, length + 1)

print(checkSet([]))
print(checkSet([1, 2, 3, 1]))
print(checkSet([1, 2, 3, 4]))

print(distinctElements([]))
print(distinctElements([1, 2, 3, 4]))
print(distinctElements([1, 2, 3, 4, 1]))
print(distinctElements([1, 2, 3, 1, 1, 2, 3, 5, 6, 5]))

File: PLF/labs/lab1/main.pl
% (i, i)
% listContains(el, l1l2...ln) = 
%   - true, el = l1
%   - false, n = 0
% 
listContains(E, [E|_]).
listContains(E, [_|L]):- listContains(E, L).

% Element, Source, dest
% (i, i, o)
% Element, list
% - [E], n = 0
% - l1 U AddToEnd(E, l2...ln), otherwise

addToEnd(E, [], [E]).
addToEnd(E, [H|T], [H|R]):- addToEnd(E, T, R).

% removeRepeated(l1l2...ln, k1k2...km) = 
%   - [], if n = 0
%   - removeRepeated(l2...ln, l1 U k1k2...km), listContains(l1, k1k2...km) = true
%   - removeRepeated(l2...ln, l1 U k1k2...km), listContains(l1, l2...ln) = true
%   - l1 U removeRepeated(l2...ln, k1k2...km), otherwise

% (i, i, i, o)
% removeRepeated - source, repeated, result, collector
removeRepeated([], _, C, C).

removeRepeated([H|T], Rep, R, C):-
    listContains(H, Rep),
    removeRepeated(T, [H|Rep], R, C).

removeRepeated([H|T], Rep, R, C):-
    listContains(H, T),
    removeRepeated(T, [H|Rep], R, C).

removeRepeated([H|T], Rep, R, C):-
    addToEnd(H, R, Res),
    removeRepeated(T, Rep, Res, C).

removeRepeatedMain(L, R):- removeRepeated(L, [], [], R).

% (i, i)
% isBiggest(el, l1l2...ln) = 
%   - true, n = 0
%   - isBiggest(el, l2...ln), el >= l1
%   - false, otherwise

isBiggest(_, []).
isBiggest(E, [H|T]):-
    E >= H,
    isBiggest(E, T).

% (i, i, i, o)
% removeBiggest(l1l2...ln, c1c2...cm) = 
%   - [], n = 0 
%   - removeBiggest(l2l3...ln, c1c2...cm), isBiggest(l1, c1c2...cm)
%   - l1 U removeBiggest(l2l3...ln, c1c2...cm), otherwise

removeBiggest([], _, C, C).

removeBiggest([H|T], CS, R, C):-
    isBiggest(H, CS),
    removeBiggest(T, CS, R, C).

removeBiggest([H|T], CS, R, C):- 
    addToEnd(H, R, Res),
    removeBiggest(T, CS, Res, C).

% (i, 0)
removeBiggestMain(L, R):- removeBiggest(L, L, [], R).